File: templates/python/v1/chain/handlers.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(slots=True)
class Request:
    id: str
    payload: str


@dataclass(slots=True)
class Handler:
    next: Optional["Handler"] = None

    def link_with(self, next_handler: "Handler") -> "Handler":
        self.next = next_handler
        return next_handler

    def handle(self, request: Request) -> str:
        result = self.process(request)
        if result is not None:
            return result
        if self.next is None:
            return f"Unhandled request: {request.id}"
        return self.next.handle(request)

    def process(self, request: Request) -> Optional[str]:
        raise NotImplementedError


class ValidationHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        if not request.payload:
            return f"Validation failed for request: {request.id}"
        return None


class AuthHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        if "token:" in request.payload:
            return None
        return f"Auth failed for request: {request.id}"


class BusinessHandler(Handler):
    def process(self, request: Request) -> Optional[str]:
        return f"Processed request {request.id} with payload length {len(request.payload)}"

File: templates/javascript/v1/chain/handlers.js
export class Handler {
  constructor() {
    this.next = null;
  }

  linkWith(next) {
    this.next = next;
    return next;
  }

  handle(request) {
    const result = this.process(request);
    if (result !== null && result !== undefined) {
      return result;
    }
    if (!this.next) {
      return `Unhandled request: ${request.id}`;
    }
    return this.next.handle(request);
  }

  process(_request) {
    throw new Error("Not implemented");
  }
}

export class ValidationHandler extends Handler {
  process(request) {
    if (!request.payload) {
      return `Validation failed for request: ${request.id}`;
    }
    return null;
  }
}

export class AuthHandler extends Handler {
  process(request) {
    if (request.payload.includes("token:")) {
      return null;
    }
    return `Auth failed for request: ${request.id}`;
  }
}

export class BusinessHandler extends Handler {
  process(request) {
    return `Processed request ${request.id} with payload length ${request.payload.length}`;
  }
}
